### PR TITLE
18681 - EFT GL Distribution supporting changes

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/18681 

*Description of changes:*
- EFTGLTransfer table to track and process EFT funds transferring between GLs (used for payment jobs for partner distribution)
- Update to existing EFT Credit to add  eft_transaction_id FK so credits can be traced back to TDI17
- Updated deduct credit logic so credit applications can be linked to an invoice (restructure to help with reconciliation and TDI17 processing traceability)
- test updates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

**Comment by ochiu** [comment]
PRs for payment-jobs and reconciliation will be submitted for this ticket as well, I need to get this one reviewed and merged for dependency updates.

File: pay-api/src/pay_api/services/eft_gl_transfer.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        if search_criteria.is_processed is not None:
+            query = query.filter(EFTGLTransferModel.is_processed == search_criteria.is_processed)
+
+        if search_criteria.processed_on:
+            query = query.filter(func.DATE(EFTGLTransferModel.processed_on) == search_criteria.processed_on.date())
+
+        if search_criteria.short_name_id:
+            query = query.filter(EFTGLTransferModel.short_name_id == search_criteria.short_name_id)
+
+        if search_criteria.source_gl:
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_gl_transfer.py:94]: Might be a good idea to write a helper function for this (could also check for date type and handle it for dates)

eg:

`query.filterIfPresent(search_criteria.source_gl, EFTGLTransferModel.short_name_id == search_criteria.short_name_id)`

File: pay-api/src/pay_api/services/payment_account.py [hunk 2]
```diff
@@ -675,15 +676,21 @@ def deduct_eft_credit(cls, invoice: InvoiceModel):
         # Deduct credits and apply to the invoice
         now = datetime.now()
         for eft_credit in eft_credits:
+            credit_invoice_link = EFTCreditInvoiceLinkModel(
+                eft_credit_id=eft_credit.id,
+                invoice_id=invoice.id
+            )
+            _ = credit_invoice_link.save() if auto_save else db.session.add(credit_invoice_link)
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:683]: helper function, probably cleaner than doing this

File: pay-api/src/pay_api/services/eft_gl_transfer.py (hunk 1 above)
Comment (seeker25) [reply pay-api/src/pay_api/services/eft_gl_transfer.py:94]: We could use this pattern in auth-api as well, where we have a bunch of conditions like below

File: pay-api/src/pay_api/services/payment_account.py (hunk 2 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/payment_account.py:683]: I was thinking the same thing, will add it to the base so we can re-use this easily

**Review by seeker25 — APPROVED** [review]
LGTM - Edit gotta fix CI

File: pay-api/src/pay_api/services/eft_gl_transfer.py [hunk 3]
```diff
@@ -76,33 +76,27 @@ def search(search_criteria: EFTGlTransferSearch = EFTGlTransferSearch()) -> [EFT
...
+        query = query.order_by(EFTGLTransferModel.created_on.asc())
 
-        if search_criteria.transfer_type:
-            query = query.filter(EFTGLTransferModel.transfer_type == search_criteria.transfer_type)
+        return query.all()
 
-        if search_criteria.transfer_date:
-            query = query.filter(func.DATE(EFTGLTransferModel.transfer_date) == search_criteria.transfer_date.date())
+    @staticmethod
+    def filterIfPresent(query, search_criteria, model_attribute):
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_gl_transfer.py:94]: This should be generic, put it in the base object?

**Review by Jxio — APPROVED** [review]
LGTM

File: pay-api/src/pay_api/services/eft_gl_transfer.py (hunk 1 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/eft_gl_transfer.py]: Updated with CustomQuery to extend the BaseQuery with filter_conditionally helper

File: pay-api/src/pay_api/services/eft_gl_transfer.py (hunk 3 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/eft_gl_transfer.py]: As discussed, extended BaseQuery with CustomQuery class that implements filter_conditionally helper

**Pull request merged**